python/whl.sh

```diff
@@ -0,0 +1,82 @@
+#!/bin/bash -e
+
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# TODO(mattmoor): Ideally we'd rewrite this in Python for portability, but
+# since the executes during WORKSPACE instantiation that would require
+# redistributing PAR files, since we cannot invoke built tools at that time.
+WHL="$1"
+REQUIREMENTS="$2"
+PKG=$(basename "${WHL}" | cut -d'-' -f 1)
+DIST_INFO="$(basename "${WHL}" | cut -d'-' -f -2).dist-info"
+
+unzip "${WHL}" > /dev/null
+
+# Enumerate the names of dependencies from the WHL metadata.
+function dependencies() {
+    cat | python <<EOF
+import json
+
+# TODO(mattmoor): find a schema / specification to follow.
+# Why are there so many layers?
+with open("${DIST_INFO}/metadata.json", "r") as f:
+  reqs = json.loads(f.read()).get("run_requires", [])
```

Comment:
In the interests of Python 2/3 compatibility, I'd suggest some minor changes:

```
import io
import json

with io.open("${DIST_INFO}/metadata.json", "r", encoding="utf8") as f:
    for req in reqs:
    requires = req.get("requires", [])
        for entry in requires:
            (name, constraint) = entry.split(" ", 1)
            print(name)
```

Also, add a TODO that we should do something about version constraints.

Reply:
This changed a little bit in `piptool.py`, it's now using `zipfile.open`.

I look forward to you beating me up on the best ways to write Python 2/3 (and ideally enforce cross-compatibility, if it's possible in Bazel).  Apologies again for the crossed signals.